sdk/metric/internal/exemplar/rand.go

```diff
@@ -0,0 +1,106 @@
+
+import (
+	"context"
+	"math"
+	"math/rand"
+	"time"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+)
+
+// rng is used to make sampling decisions.
+//
+// Do not use crypto/rand. There is no reason for the decrease in performance
+// given this is not a security sensitive decision.
+var rng = rand.New(rand.NewSource(time.Now().UnixNano()))
+
+// FixedSize returns a [Reservoir] that samples at most n exemplars. If there
+// are n or less measurements made, the Reservoir will sample each one. If
+// there are more than n, the Reservoir will then randomly sample all
+// additional measurement with a decreasing probability.
+func FixedSize[N int64 | float64](n int) Reservoir[N] {
+	r := &randRes[N]{storage: newStorage[N](n)}
+	r.reset()
+	return r
+}
+
+type randRes[N int64 | float64] struct {
+	*storage[N]
+
+	// count is the number of measurement seen.
+	count int64
+	// next is the next count that will store a measurement at a randon index
+	// once the reservoir has been filled.
+	next int64
+	// w is the largest random number in a distribution that is used to compute
+	// the next next.
+	w float64
+}
+
+func (r *randRes[N]) Offer(ctx context.Context, t time.Time, n N, a []attribute.KeyValue) {
+	// The following algorithm is "Algorithm L" from Li, Kim-Hung (4 December
+	// 1994). "Reservoir-Sampling Algorithms of Time Complexity
+	// O(n(1+log(N/n)))". ACM Transactions on Mathematical Software. 20 (4):
+	// 481–493 (https://dl.acm.org/doi/10.1145/198429.198435).
+	//
+	// It is used because of its balance of simplicity and performance. In
+	// particular it has an asymptotic runtime of O(k(1 + log(n/k)) where n is
+	// the number of measurements offered and k is the reservoir size. This is
+	// much more optimal for large measurement sets than the algorithm
+	// recommended by the OTel spcification ("Algorithm R" as described in
+	// Vitter, Jeffrey S. (1 March 1985). "Random sampling with a reservoir"
+	// (http://www.cs.umd.edu/~samir/498/vitter.pdf)) which has an asymptotic
+	// runtime of O(n).
+	//
+	// See https://github.com/MrAlias/reservoir-sampling for a comparison of
+	// reservoir sampling algorithms (including performance benchmarks).
+
+	if int(r.count) < cap(r.store) {
+		r.store[r.count] = newMeasurement(ctx, t, n, a)
+	} else {
+		if r.count == r.next {
+			idx := int(rng.Int63n(int64(cap(r.store))))
+			r.store[idx] = newMeasurement(ctx, t, n, a)
+			r.advance()
+		}
+	}
+	r.count++
+}
+
+func (r *randRes[N]) reset() {
+	r.count = 0
+	r.next = int64(cap(r.store))
+	r.w = math.Exp(math.Log(rng.Float64()) / float64(cap(r.store)))
+	r.advance()
+}
+
+func (r *randRes[N]) advance() {
+	r.w *= math.Exp(math.Log(rng.Float64()) / float64(cap(r.store)))
+	r.next += int64(math.Log(rng.Float64())/math.Log(1-r.w)) + 1
+}
+
+func (r *randRes[N]) Collect(dest *[]metricdata.Exemplar[N]) {
+	r.storage.Collect(dest)
+	r.reset()
```

Comment:
If collect isn't supposed to clear data, I wouldn't expect it to reset here.  Won't it just immediately overwrite all of the existing exemplars then, making it roughly equivalent to Flush?

Reply:
This comes from the [specification](https://github.com/MrAlias/opentelemetry-specification/blob/main/specification/metrics/sdk.md#simplefixedsizeexemplarreservoir):

> Any stateful portion of sampling computation SHOULD be reset every collection
> cycle. For the above example, that would mean that the `num_measurements_seen`
> count is reset every time the reservoir is collected.

Your are correct in that if more measurements are made in the next collection cycle this will be equivalent to a flush. But I think the specification was trying to handle two things:

- If no measurements are made in the next collection cycle persist old exemplars
- If there are measurements made in the next collection cycle, prioritize those as exemplars.

In the end this method will be different from `Flush` in that it will always return `n` exemplars once `n` exemplars have been recorded.

Reply:
commented: https://github.com/open-telemetry/opentelemetry-go/pull/4852/commits/b8af1e7dccea17c8af203133f0abe85ecb8b4194

Reply:
thanks.  Makes sense.